lwgenerate/azure/azure.go

```diff
@@ -0,0 +1,458 @@
+		),
+		lwgenerate.NewRequiredProvider(
+			"azurerm",
+			lwgenerate.HclRequiredProviderWithSource(lwgenerate.HashAzureRMProviderSource),
+			lwgenerate.HclRequiredProviderWithVersion(lwgenerate.HashAzureRMProviderVersion),
+		),
+	)
+}
+
+func createAzureADProvider() ([]*hclwrite.Block, error) {
+	blocks := []*hclwrite.Block{}
+	//attrs := map[string]interface{}{}
+	provider, err := lwgenerate.NewProvider(
+		"azuread",
+	).ToBlock()
+
+	if err != nil {
+		return nil, err
+	}
+
+	blocks = append(blocks, provider)
+	return blocks, nil
+}
+
+// In this we need to create a provider block with a  features
+// configuration but with nothing set,  this is as per the
+// Azure examples and is of the format
+//
+//         provider "azurerm" {
+//            features = {}
+//         }
+//
+func createAzureRMProvider() ([]*hclwrite.Block, error) {
+	blocks := []*hclwrite.Block{}
+	attrs := map[string]interface{}{}
+
+	provider, err := lwgenerate.NewProvider(
+		"azurerm",
+	).ToBlock()
+
+	if err != nil {
+		return nil, err
+	}
+	// Create the features block
+	featuresBlock, err := lwgenerate.HclCreateGenericBlock("features", []string{}, attrs)
+	provider.Body().AppendBlock(featuresBlock)
+
+	if err != nil {
+		return nil, err
+	}
+
+	blocks = append(blocks, provider)
+	return blocks, nil
+}
+
+func createLaceworkAzureADModule(args *GenerateAzureTfConfigurationArgs) ([]*hclwrite.Block, error) {
+	blocks := []*hclwrite.Block{}
+
+	if args.CreateAdIntegration {
+		provider, err := lwgenerate.NewModule(
+			"az_ad_application",
+			lwgenerate.LWAzureADSource,
+			lwgenerate.HclModuleWithVersion(lwgenerate.LWAzureADVersion),
+		).ToBlock()
+
+		if err != nil {
+			return nil, err
+		}
+
+		blocks = append(blocks, provider)
+	}
+	return blocks, nil
+}
+
+func createConfig(args *GenerateAzureTfConfigurationArgs) ([]*hclwrite.Block, error) {
+	blocks := []*hclwrite.Block{}
+	if args.Config {
+		attributes := map[string]interface{}{}
+		moduleDetails := []lwgenerate.HclModuleModifier{}
+
+		if args.ConfigIntegrationName != "" {
+			attributes["lacework_integration_name"] = args.ConfigIntegrationName
+		}
+
+		// Check if we have created an Active Directory app
+		if args.CreateAdIntegration {
+			attributes["use_existing_ad_application"] = true
+			attributes["application_id"] = "module.az_ad_application.application_id"
+			attributes["application_password"] = "module.az_ad_application.application_password"
+			attributes["service_principal_id"] = "module.az_ad_application.service_principal_id"
+		} else {
+			attributes["use_existing_ad_application"] = true
+			attributes["application_id"] = args.AdApplicationId
+			attributes["application_password"] = args.AdApplicationPassword
+			attributes["service_principal_id"] = args.AdServicePrincipalId
+		}
+
+		// Only set subscription ids if all subscriptions flag is not set
+		if !args.AllSubscriptions {
+			if len(args.SubscriptionIds) > 0 {
+				attributes["subscription_ids"] = args.SubscriptionIds
+			}
```

Comment:
should we raise an error here if `!args.AllSubscriptions && len(args.SubscriptionIds) == 0`

Reply:
perhaps this block could be something like this:
```
if args.AllSubscriptions {
    // Set Subscription information
    attributes["all_subscriptions"] = args.AllSubscriptions
} else if len(args.SubscriptionIds) > 0 {
    attributes["subscription_ids"] = args.SubscriptionIds
} else {
    return an error here....
}
```

Reply:
Don't think that this works here, we seem to return an error in the case where `AllSubscriptions` is set to `false`

If `AllSubscriptions` is set `true`, then it does not matter what is in the `SubscriptionIds` list, so we only need to look at this list in the case where it is `false`

Reply:
I wouldn't expect that to be the behavior here. If all subscriptions is false, it will check if we have any subscription id's in the else if. Only if both all subscriptions is false and subscriptions id is empty should we hit the else and return an error.

Reply:
If all subscriptions is true it would just set that to true and skip the else if and else, right?

Reply:
OK I see what you are getting at now, misunderstanding on my part. The thing is that we do not really need the else part. 
If `AllSubscriptions` is set `true`, it will use all and we only need to set the flag
If `AllSubscriptions` is set `false`, we may or may not have a list of `SubscriptionIds` , if we have we set them and they are used, if there are none then it will default to using the primary subscription for the account.

So in my mind we have the simpler logic that is in the original I think

Reply:
Fair enough. I was unaware that we had that fall-through option on the subscription id. In that case we're good as is